app/models/account.rb

```diff
@@ -36,18 +40,56 @@ class Account < ApplicationRecord
   # @return [Account] A persisted instance of Account
   #
   def self.cache_or_create_by!(user)
-    return unless user.uuid
+    return unless user.uuid || user.sec_id
 
-    do_cached_with(key: user.uuid) do
+    # if possible use the idme uuid for the key, fallback to using the sec id otherwise
+    key = user.uuid || "sec:#{user.sec_id}"
+    acct = do_cached_with(key: key) do
       create_if_needed!(user)
     end
+    # Account.sec_id was added months after this class was built, thus
+    # the existing Account records (not new ones) need to have their
+    # sec_id value updated
+    update_if_needed!(acct, user)
   end
 
   def self.create_if_needed!(user)
-    find_or_create_by!(idme_uuid: user.uuid) do |account|
-      account.edipi = user&.edipi
-      account.icn   = user&.icn
+    attrs = account_attrs_from_user(user)
+
+    accts = where(idme_uuid: attrs[:idme_uuid])
+            .where.not(idme_uuid: nil)
+            .or(
+              where(sec_id: attrs[:sec_id])
+              .where.not(sec_id: nil)
+            )
+
+    if accts.length > 1
+      data = accts.map(&:attributes)
+      log_message_to_sentry('multiple Account records with matching ids', 'warning', data)
     end
```

Comment:
Could we instead create a custom uniqueness constraint on this model requiring that (`idme_uuid `+`sec_id`) must be  unique?

Also, is it ok to just continue marching on after finding more than 1 account? Some other options would be raising, or trying to consolidate the accounts into 1.

Reply:
I think we're okay with constraints where we are. For a duplicate idme_uuid to occur, MVI would have to have mapped an identity on to different individuals at different times, associating the idme_uuid with two different secids. We should definitely log that to sentry and analyze if it does happen but I don't think it's likely, and I don't think the custom uniqueness constraint express much of use. 

If we do find more than one account, you're right that we probably need a heuristic to pick one other than the unstable ordering coming back from the database, but I'd also like to observe this and see if sentry picks up any of these events.